README.md
[![Build status](https://ci.appveyor.com/api/projects/status/ymdjaaa9nk63gsm2?svg=true)](https://ci.appveyor.com/project/fusioninventory/fusioninventory-agent-windows-installer)
# FusionInventory Agent Windows Installer

## Description

The FusionInventory Agent Windows Installer is an open source project with
as goal to build the installer program of the FusionInventory Agent and its tasks
for Windows operative systems. It makes use of some others open source projects,
like the Nullsoft Scriptable Install System (in forward NSIS), Strawberry Perl,
Curl, etcetera.

It was born to cover a basic goal; be able to install new FusionInventory Agent
versions based on the previous configuration in the system, whether it exists.
In summary, it is born to be able to update the existing agent, and not only to
install a new version from scratch.

It has more purposes, of course. The following are some of them.

## Features

   - Installs from scratch or from the current configuration.

   - Uninstalls the previous agent, whether it exists.

   - Prevents multiple installations simultaneously.

   - Supports visual installation in multiple languages.
        (English and Spanish. French in construction.)

   - Builds two different installers for x86 and x64 architectures.
        (Each platform uses its native distribution of Strawberry Perl)

   - Builds installers for stable or development versions.

   - Supports both installation methods; silent or unattended mode and
     graphical or visual mode.

   - Allows to select the agent tasks to install.

   - New visual appearance based on the NSIS Modern UI 2 plugin.

   - Migrates the deprecated options to the new options and removes completely
     the obsolete ones from the Microsoft Windows registry.

   - Now the Microsoft Windows registry used for agent configuration integrates
     all the options supported by the agent, and not only those that them
     values are different to the default.

   - Allows a complete customization of all the options supported for the
     agent, either from the the command line, or from the visual installation.

   - Each generated installer is identified uniquely by a BuildID.
        (Each architecture has its own BuildID sequence)

   - Allows to execute the agent as a Windows Service, to plan its execution
     through a Windows Task or, simply, not to execute the agent.

   - Allows to pull a SSL certificate from a URL at installation time. (ToDo)

## Changes

See [Changes.txt file](Changes.txt)

## Dependencies

Softwares used for building:

 * [NSIS](http://nsis.sourceforge.net/Download) 3.02
 * Pandoc 1.17.0.2

## Release process

 * For a release:
    1. Set a tag matching `2.4(.1)` scheme on [fusioninventory/fusioninventory-agent](https://github.com/fusioninventory/fusioninventory-agent) repository
    1. Set the same tag on this repository
    1. Release are then available in [this fusioninventory repository releases](https://github.com/fusioninventory/fusioninventory-agent-windows-installer/releases)

 * For any development release:
    1. Update agent code if necessary on [fusioninventory/fusioninventory-agent](https://github.com/fusioninventory/fusioninventory-agent) repository default branch
    1. Update [Changes.txt](Changes.txt) as needed, keeping agent commit ref in comment
    1. Development builds are then available at [built artifacts page](https://ci.appveyor.com/project/fusioninventory/fusioninventory-agent-windows-installer/build/artifacts)

## Manual build

See [Legacy build section in Readme.txt file](Readme.txt)

## Contacts

FusionInventory project websites:

* main site: <http://www.fusioninventory.org>
* contact ways: <http://fusioninventory.org/resources/>

## Bug reporting

Please, follow [our bug reporting documetation](http://fusioninventory.org/documentation/bugreport/) to report any issues.

## License

This software is licensed under the terms of GPLv2, see [License.txt file](License.txt) file for
details.


Changes.txt
FusionInventory Agent Windows Installer - History of Revisions and Changes

2.6 Thu, 26 Nov 2020
agent:
* FusionInventory Agent 2.6 (commit: e4192a2 | date: nov 26 2020)
* as explain in Version.pm comments, actual versioning is problematic for CPAN versions,
  so any future version will probably be defined as only major.minor but evantually
  as major.minor_rev if a package revision is necessary. This policy could be revised
  after any major upgrade.
* Fix configuration reloading to keep set script options
* Avoid a not critical perl error while reloading HTTP server
* Fix #771: added Acer monitor serial support for models (G227HQL, G236HL, R221Q, S273HL)
* Battery capacity support provides now canonical values
* Battery capacity support permits to compute battery capacity usage, thanks to Thierry Bugier
* Fix #819: Surface Go 2 was recongized has Hyper-V
* Fix #829: added other Acer monitor serial support (H226HQL, K222HQL, SA220Q)
* Dmidecode output parsing cleanup
* Better canonical manufacturer reporting
* Added PartNumber API to enhanced inventory memory and controllers support:
 * Elpida, Hynix, Micron & Samsung memory partnumbers
 * Dell controller partnumbers
* thanks to Vadim Pisarev:
 * Memory inventory includes partnumber as MODEL
 * Controller inventory includes SERIAL and MODEL
 * Micron memories PartNumber support
* Updated Ricoh printers support with better hostname detection
* Updated Zebra printers support with better hostname detection
* Introduced Siemens industrial modules support
* Introduced Multitech industrial modules support
* Fix #823, #856: Add Ubiquiti UniFi AP linux appliance support
* Fix #772: Enhanced some HP switchs support
* Introduced Hwg support, thanks to Bertrand Keller
* injector: Fix an issue with --xml-ua option triggering a perl fatal error on empty or wrong XML files

core:
* Portable version don't rely on registry config backend by default
* Update PCI IDs database (pci.ids file) to 2019.12.11
* Update USB IDs database (usb.ids file) to 2019.11.05

strawberry perl package:
* Based on Strawberry Perl 5.32.0.1 (August 2020) revision 1

tools:
* Updated dmidecode to v3.3-gitd64af97

2.5.2 Mon, 16 Dec 2019
agent:
* FusionInventory Agent 2.5.2 (commit: 7f6ee84481 | date: dec 16 2019)
* fix HTTP server IPv6
* win32: always detach agent thread after 10 seconds when stopping the agent
  service when it is blocking on anything to guaranty the service stops in 10s max
* check agent persistent datas for a "forcerun" set flag. This is firstly intended
  to be used by win32 installer to handle the "start inventory after installation"
  option directly from the service.
  This can also be used to change or reset the agent deviceid.
* Add --set-forcerun option support to fusioninventory-agent script
* fix: don't reschedule too early on a forced run at start
* fix: Don't use delaytime on config reload
* fix: use target counters reset to better support config reload
* feature: limit next run delay reduction
  Limit next run random delay reduction to max 1/6 of the delay if less than 6 hours,
  limit to max an hour for delay from 6 hours to 24 hours,
  and limit to max 1/24 of the delay for delay greater than a day.
  This would keep enough delay randomization to avoid mass agent server connection and
  keep next seen run more coherent with the requested delay.
* fix service shutdown when HTTP client close the connection before the agent
* fix HTTP server keep-alive support with a 8 requests by connection maximum limit
* service update to support a safe forking system to firstly support handling
  parallel http request for the Proxy HTTP server plugin.
* Proxy HTTP server plugin now support max_proxy_threads configuration which is
  set to 10 maximum concurrent requests by default.
* added Samsung monitor serial support for models: B1940MR, B1940W, S22A450BW,
  S22B420, S22E450, S22F350FHU, S27D390H, S27D850T, S27H850QFU, S19A450, SM943BM,
  S22C450, S27H650FDU
* win32: fix hyperv host wasn't filtered from installed virtual machines
* win32: fix Adobe key detection, thanks to PR-gh
* win32: fix firewall inventory cases, thanks to PR-gh
* win32: fix Office license inventory, thanks to PR-gh
* win32: enhanced network card inventory, thanks to PR-gh
* win32: fix software inventory failing due to unsupported UTF-16 Appx manifest XML
* fix #359: try to obtain remote management litemanager ID under win32
* Add Tagged VLAN, work at least with HPE Switch, thanks to PR-gh
* Enhanced Qlogic support to report device serial number, thanks to po1vo
* Added StormShield support via FreeBSD MIBSupport, thanks to PR-gh
* Fix memory and storage size normalization in some cases, thanks to PR-gh
* fix #738: fix time expiration when netinventory is chained with netdiscovery
  while using fusioninventory-netdiscovery script --inventory option
* fix #741: device with empty description was no more inventoried
* fix #726: Added support for latest SPARC cpus on Solaris
* fix #717: use cdpCacheSysName when available to enhance connection detection
  Try also to extract remote mac address from deviceId when possible (fix Meraki
  connection detection, even support Yealink SIP phones)
* fix #684: Added IAP Aruba serial number support
* moved APC pdu support to mibsupport, thanks to po1vo
* fix #751: Added Seagate storage support
* enhanced VLAN support
* fix #734: Fix issue with Ricoh printers
* fix #722: Enhance LinuxAppliance support with Sophos UTG support
* improved Ricoh printer support
* Updated sysobject.ids (tagged fia-2.5.2 tag on github repository)

core:
* /runnow involves now to force a run at next service or command start
* Installer supports now a portable install mode, it is used to build the
  portable package
* Don't forget to install HTTP server plugin default configurations
* Update PCI IDs database (pci.ids file) to 2019.12.11
* Update USB IDs database (usb.ids file) to 2019.11.05

strawberry perl package:
* Based on Strawberry Perl 5.30.0.1 (May 2019) revision 2

NSIS/FusionInventory-Agent-Portable.sh
#!/bin/bash
# ------------------------------------------------------------------------
# FusionInventory Agent Installer for Microsoft Windows
# Copyright (C) 2010-2019 by the FusionInventory Development Team.
#
# http://www.fusioninventory.org/
# ------------------------------------------------------------------------
#
# LICENSE
#
# This file is part of FusionInventory project.
#
# FusionInventory Agent Installer for Microsoft Windows is free software;
# you can redistribute it and/or modify it under the terms of the GNU
# General Public License as published by the Free Software Foundation;
# either version 2 of the License, or (at your option) any later version.
#
#
# FusionInventory Agent Installer for Microsoft Windows is distributed in
# the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
# the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
# PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA,
# or see <http://www.gnu.org/licenses/>.
#
# ------------------------------------------------------------------------
#
# @package   FusionInventory Agent Installer for Microsoft Windows
# @file      .\NSIS\FusionInventory-Agent-Portable.sh
# @author    Tomas Abad <tabadgp@gmail.com>
#            Guillaume Bougard <gbougard at teclib.com>
# @copyright Copyright (c) 2010-2019 FusionInventory Team
# @license   GNU GPL version 2 or (at your option) any later version
#            http://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# @link      http://www.fusioninventory.org/
# @link      https://github.com/fusioninventory/fusioninventory-agent
# @since     2012
#
# ------------------------------------------------------------------------

declare -r installer="$1"

# Load perl environment
source ../Perl/Scripts/load-perl-environment

declare arch=''
declare digest=''
declare basename=''
declare -a -r digests=(md5 sha1 sha256)

declare -r openssl=$(type -P openssl)

# Check the OS
if [ "${MSYSTEM}" = "MSYS" ]; then
   # Windows OS with MinGW/MSYS

   basename="${0##*\\}"
else
   if [ -n "${WINDIR}" ]; then
      # It's a Windows OS

      basename="${0##*\\}"

      echo
      echo -n "You can not launch '${basename}' directly. "
      echo "Please, launch '${basename%.sh}.bat' instead."
      echo

      exit 1
   fi

   # It's a UNIX OS.

   basename="${0##*/}"
fi

if [ -z "$installer" -o ! -e "$installer" ]; then
   basename="${0##*\\}"

   echo
   echo -n "You can not launch '${basename}' directly without an installer as argument. "
   echo "Please, launch '${basename%.sh}.bat' instead."
   echo

   exit 2
fi

if [ ! -d "Portable/FusionInventory-Agent" ]; then
   basename="${0##*\\}"

   echo
   echo "Agent not installed at the expected place"
   echo
   echo "Please, launch '${basename%.sh}.bat' only after installers has been built."
   echo

   exit 3
fi

# All seems be correct...

# Extract arch from installer name
arch=${installer##*windows-}
arch=${arch%_*}

# Build installer related portable archive
echo -n "Building ${arch} portable archive..."

# Add data dir
/bin/install --directory "Portable/FusionInventory-Agent/data"

# Unset logfile path in portable.cfg to disable its usage by default
/bin/sed -i -e "s|^logfile.*$|logfile =|" "Portable/FusionInventory-Agent/etc/conf.d/portable.cfg"

# Portable version should not use registry as default backend, this sed comments the l.105 in Config.pm
/bin/sed -i -e "s|^        \$OSNAME|        #\$OSNAME|" "Portable/FusionInventory-Agent/perl/agent/FusionInventory/Agent/Config.pm"

( cd Portable ; 7z a -bd -sfx7z.sfx -stl -y "../${installer%.exe}-portable.exe" "FusionInventory-Agent" >../7z-${arch}-portable.txt 2>&1; )
if (( $? == 0 )); then
   echo '.Done!'

   # Digest calculation loop
   echo -n "Calculating digest message for ${arch} portable archive."
   for digest in "${digests[@]}"; do
      "${openssl}" dgst -${digest} -c -out "${installer%.exe}-portable.${digest}" "${installer%.exe}-portable.exe"
      echo -n "."
   done
   echo ".Done!"
else
   echo '.Failure!'
   echo " Failed to build ${arch} portable agent archive."
fi

/bin/rm -rf "Portable" > /dev/null 2>&1

echo


Perl/Scripts/patch-fusioninventory-agent.sh
#!/bin/bash
# ------------------------------------------------------------------------
# FusionInventory
# Copyright (C) 2010-2013 by the FusionInventory Development Team.
#
# http://www.fusioninventory.org/   http://forge.fusioninventory.org/
# ------------------------------------------------------------------------
#
# LICENSE
#
# This file is part of FusionInventory project.
#
# FusionInventory Agent Installer for Microsoft Windows is free software;
# you can redistribute it and/or modify it under the terms of the GNU
# General Public License as published by the Free Software Foundation;
# either version 2 of the License, or (at your option) any later version.
#
#
# FusionInventory Agent Installer for Microsoft Windows is distributed in
# the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
# the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
# PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA,
# or see <http://www.gnu.org/licenses/>.
#
# ------------------------------------------------------------------------
#
# @package   FusionInventory Agent Installer for Microsoft Windows
# @file      .\Perl\Scripts\patch-fusioninventory-agent.sh
# @author    Tomas Abad <tabadgp@gmail.com>
# @copyright Copyright (c) 2010-2013 FusionInventory Team
# @license   GNU GPL version 2 or (at your option) any later version
#            http://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# @link      http://www.fusioninventory.org/
# @link      http://forge.fusioninventory.org/projects/fusioninventory-agent
# @since     2012
#
# ------------------------------------------------------------------------


# Load perl environment
source ./load-perl-environment

declare -i iter=0
declare basename=''
declare base_path=''
declare script_suffix=''

declare -r cat=$(type -P cat)
declare -r tr=$(type -P tr)

# Check the OS
if [ "${MSYSTEM}" = "MSYS" ]; then
   # Windows OS with MinGW/MSYS

   basename="${0##*\\}"
   script_suffix='bat'
else
   if [ -n "${WINDIR}" ]; then
      # It's a Windows OS

      basename="${0##*\\}"

      echo
      echo -n "You can not launch '${basename}' directly. "
      echo "Please, launch '${basename%.sh}.bat' instead."
      echo

      exit 1
   fi

   # It's a UNIX OS.

   basename="${0##*/}"
   script_suffix='sh'
fi

# Check whether Strawberry Perl ${strawberry_path} is already installed
if [ ! -d "${strawberry_path}" ]; then
   echo
   echo "Sorry but it seems that Strawberry Perl ${strawberry_release} (${strawberry_version}-32/64bits)"
   echo "is not installed into the '${strawberry_path}' directory."
   if [ "${MSYSTEM}" = "MSYS" ]; then
      # Windows OS with MinGW/MSYS
      echo "Please, install it with 'install-strawberry-perl.bat' or with"
      echo "'install-strawberry-perl-package-for-fusioninventory-agent.sh', and try again."
   else
      # UNIX OS
      echo "Please, install it with 'install-strawberry-perl-package-for-fusioninventory-agent.sh',"
      echo "and try it again."
   fi
   echo

   exit 3
fi

# Setup loop
while (( ${iter} < ${#archs[@]} )); do
   # Set arch and arch_label
   arch=${archs[${iter}]}
   arch_label=${arch_labels[${iter}]}

   # Get base_path
   eval base_path="${strawberry_arch_path}/cpan/sources/${fusinv_agent_mod_name}-${fusinv_agent_commit}"

   # Check whether 'install-fusioninventory-agent.${script_suffix}' was launched
   if [ ! -d "${base_path}" ]; then
      echo
      echo "Sorry but it seems that you didn't install FusionInventory-Agent ${fusinv_agent_commit} before."
      echo "Please, install it with 'install-fusioninventory-agent.${script_suffix}' and try again."
      echo

      break
   fi

   # Display task
   echo -n "Setting up FusionInventory-Agent ${fusinv_agent_commit} for Strawberry Perl ${strawberry_release} (${strawberry_version}-${arch_label}s)."

   # Updated version release ${base_path}/lib/FusionInventory/Agent/Version.pm
   # with version tag
   if [ -n "${fusinv_agent_release}" ]; then
      PACKAGE_TIME=$(LANG=C date -u)

      FIA_VERSION="${fusinv_agent_release}"
      if [ "${TYPE}" = "development" ]; then
         BUILD=${APPVEYOR_BUILD_NUMBER}
         : ${BUILD:=0000}
         if [ -z "${fusinv_agent_release%%*-*}" ]; then
            FIA_VERSION="${fusinv_agent_release%%-*}-develop-${BUILD}-${fusinv_agent_release#*-}"
         else
            FIA_VERSION="${fusinv_agent_release}-develop-${BUILD}"
         fi
      fi
      : ${APPVEYOR_ACCOUNT_NAME:=$USERNAME}
      APPVEYOR_ACCOUNT_NAME="$( echo -n ${APPVEYOR_ACCOUNT_NAME:0:1} | ${tr} 'a-z' 'A-Z' )${APPVEYOR_ACCOUNT_NAME#?}"

      ${cat} >"${base_path}/lib/FusionInventory/Agent/Version.pm" <<HERE_VERSION
package FusionInventory::Agent::Version;

use strict;
use warnings;

our \$VERSION = "${FIA_VERSION}";
our \$PROVIDER = "FusionInventory";
our \$COMMENTS = [
    "Provided by ${APPVEYOR_ACCOUNT_NAME}",
    "Installer built with Appveyor on $PACKAGE_TIME"
];

1;
HERE_VERSION

      echo -n "."
   fi

   echo ".Done!"

   # New architecture
   iter=$(( ${iter} + 1 ))
done

echo
